tests/components/dsmr_reader/test_config_flow.py

```diff
@@ -32,6 +53,10 @@ async def test_user_step_with_mqtt(hass: HomeAssistant):
     assert result["step_id"] == "confirm"
     assert result["errors"] is None
 
+    # configure bogus mqtt service to pass flow
+    hass.services.async_register("mqtt", "publish", None)
+    hass.data[DATA_MQTT] = {"async_subscribe": MagicMock()}
```

Comment:
I think it would be better to check if `mqtt` is in `hass.config.components` since `publish` is an implementation detail of `mqtt` that could get renamed in the future.

```
@core.callback
def async_get_loaded_integrations(hass: core.HomeAssistant) -> set[str]:
    """Return the complete list of loaded integrations."""
    integrations = set()
    for component in hass.config.components:
        if "." not in component:
            integrations.add(component)
            continue
        domain, platform = component.split(".", 1)
        if domain in BASE_PLATFORMS:
            integrations.add(platform)
    return integrations

```

Reply:
While it seems very unlikely that the publish operation will ever change its name in the context of MQTT, I certainly see your point. The issue is just that `hass.config.components` always seems to contain at least `{'http', 'mqtt'}`, no matter what I do.

It seems like it can be fixed by checking the config entries instead, like with `return "mqtt" in hass.config_entries.async_domains()`. That will tell whether MQTT is configured. I've tested this locally and couldn't find any issues with it.